pennylane/devices/qubit/simulate.py

```diff
@@ -55,6 +60,27 @@ def simulate(circuit: qml.tape.QuantumScript) -> Union[tuple, TensorLike]:
     for op in circuit._ops:
         state = apply_operation(op, state)
 
+    if circuit.shots.total_shots is None:
+        # analytic case
+
+        if len(circuit.measurements) == 1:
+            return measure(circuit.measurements[0], state)
+
+        return tuple(measure(mp, state) for mp in circuit.measurements)
+
+    # finite-shot case
+
     if len(circuit.measurements) == 1:
-        return measure(circuit.measurements[0], state)
-    return tuple(measure(mp, state) for mp in circuit.measurements)
+        return measure_with_samples(circuit.measurements[0], state, shots=circuit.shots, rng=rng)
+
+    rng = np.random.default_rng(rng)
+    results = tuple(
+        measure_with_samples(mp, state, shots=circuit.shots, rng=rng) for mp in circuit.measurements
```

Comment:
I'm a bit concerned about performance with this implementation. Let's say that there are 1000 shots with 5 different measurements. In that case, 1000 samples will be generated each time for each measurement totalling 5000 samples. That makes a lot of sense to do in terms of keeping behaviour consistent with hardware.

However, to optimize runtime, we could instead just create 1000 samples, and use the same samples for each measurement. That's what `default.qubit` does.

Reply:
Good point. Looking at it again, I think it would be better to change the signature of `measure_with_samples` to take samples instead of the state, and just call `sample_state` once in `simulate`. @albi3ro Any thoughts on this?

Reply:
For anything that would affect user-facing workflows (e.g., results/numerics or performance), don't forget to always loop in @trbromley :)

Reply:
On second thought, this kind of goes hand-in-hand with the optimizations we mentioned before on calling `sample_state` just once. So we can probably leave this for now and come back to it when we do those optimizations.

Reply:
Right now we want to get basic behavior so we can unblock other work.  This is all implementation details, and we have lots of improvements and optimizations to do in the future.